pkg/controller/job/indexed_job_utils.go

```diff
@@ -38,82 +38,173 @@ func isIndexedJob(job *batch.Job) bool {
 	return job.Spec.CompletionMode != nil && *job.Spec.CompletionMode == batch.IndexedCompletion
 }
 
-// calculateSucceededIndexes returns a string representation of the list of
-// succeeded indexes in compressed format and the number of succeeded indexes.
-func calculateSucceededIndexes(pods []*v1.Pod) (string, int32) {
-	sort.Sort(byCompletionIndex(pods))
-	var result strings.Builder
-	var lastSucceeded int
-	var count int32
-	firstSucceeded := math.MinInt32
+type interval struct {
+	First int
+	Last  int
+}
+
+type orderedIntervals []interval
+
+// calculateSucceededIndexes returns the list of succeeded indexes in
+// in compressed format (intervals).
+// The first list only contains the indexes in .status.completedIndexes if
+// tracking with finalizers. The second list includes the indexes that
+// succeeded since the last sync.
```

Comment:
Maybe just say it returns before and after intervals for ease of creating a patch?